### PR TITLE
GH-111848: Convert remaining jumps to deopts into tier 2 code.

### PR DESCRIPTION
**Pull request by markshannon**
* Issue: gh-111848

### CODE REVIEW

File: Python/optimizer.c [hunk 1]
```diff
@@ -385,14 +385,26 @@ PyTypeObject _PyUOpExecutor_Type = {
...
+static const uint16_t
+BRANCH_TO_GUARDS[4][2] = {
+    [POP_JUMP_IF_FALSE - POP_JUMP_IF_FALSE][0] = _GUARD_IS_TRUE_POP,
+    [POP_JUMP_IF_FALSE - POP_JUMP_IF_FALSE][1] = _GUARD_IS_FALSE_POP,
+    [POP_JUMP_IF_TRUE - POP_JUMP_IF_FALSE][0] = _GUARD_IS_FALSE_POP,
+    [POP_JUMP_IF_TRUE - POP_JUMP_IF_FALSE][1] = _GUARD_IS_TRUE_POP,
+    [POP_JUMP_IF_NONE - POP_JUMP_IF_FALSE][0] = _GUARD_IS_NOT_NONE_POP,
+    [POP_JUMP_IF_NONE - POP_JUMP_IF_FALSE][1] = _GUARD_IS_NONE_POP,
+    [POP_JUMP_IF_NOT_NONE - POP_JUMP_IF_FALSE][0] = _GUARD_IS_NONE_POP,
+    [POP_JUMP_IF_NOT_NONE - POP_JUMP_IF_FALSE][1] = _GUARD_IS_NOT_NONE_POP,
```
Comment (iritkatriel) [inline Python/optimizer.c:405]: What is supposed to guarantee that the jump opcode ids are consecutive? Their lexicographer order?

Comment (markshannon) [reply Python/optimizer.c:405]: The C compiler ensures this is correct, at least gcc does. 
"array index in initializer exceeds array bounds" if any of the indices exceed the bounds.

We should be generating these tables, but the code generator needs refactoring first.

File: Python/optimizer.c [hunk 2]
```diff
@@ -385,14 +385,26 @@ PyTypeObject _PyUOpExecutor_Type = {
...
+/* TO DO -- Generate these tables */
 static const uint16_t
 _PyUop_Replacements[OPCODE_METADATA_SIZE] = {
     [_ITER_JUMP_RANGE] = _GUARD_NOT_EXHAUSTED_RANGE,
     [_ITER_JUMP_LIST] = _GUARD_NOT_EXHAUSTED_LIST,
     [_ITER_JUMP_TUPLE] = _GUARD_NOT_EXHAUSTED_TUPLE,
 };
 
+static const uint16_t
+BRANCH_TO_GUARDS[4][2] = {
```
Comment (iritkatriel) [inline Python/optimizer.c]:
```suggestion
BRANCH_TO_GUARD[4][2] = {
```

File: Python/optimizer.c [hunk 3]
```diff
@@ -776,6 +766,16 @@ translate_bytecode_to_trace(
 #define SET_BIT(array, bit) (array[(bit)>>5] |= (1<<((bit)&31)))
 #define BIT_IS_SET(array, bit) (array[(bit)>>5] & (1<<((bit)&31)))
 
+static bool
+is_branch(int opcode) {
```
Comment (iritkatriel) [inline Python/optimizer.c]: There's `OPCODE_HAS_JUMP`

**Review by iritkatriel — APPROVED** [review]

**Pull request merged**